### PR TITLE
channeldb+lnrpc: invoice state

### PR DESCRIPTION
**Pull request by joostjager**
This PR converts the invoice `settled` boolean into a state. This is a preparation that enables creation of new invoice states.

### CODE REVIEW

File: channeldb/invoices.go [hunk 1]
```diff
@@ -74,6 +74,18 @@ const (
 	MaxPaymentRequestSize = 4096
 )
 
+// ContractState describes the state the invoice is in.
+type ContractState uint8
+
+const (
+	// ContractOpen means the invoice has only been created.
+	ContractOpen = 0
```
Comment (halseth) [inline channeldb/invoices.go]: should be set to be of type `ContractState`

File: htlcswitch/link.go [hunk 2]
```diff
@@ -2334,7 +2334,7 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
 			// TODO(conner): track ownership of settlements to
 			// properly recover from failures? or add batch invoice
 			// settlement
-			if invoice.Terms.Settled {
+			if invoice.Terms.State != channeldb.ContractOpen {
```
Comment (halseth) [inline htlcswitch/link.go]: Is this intended? Comment should possibly be updated to reflect change?

File: invoiceregistry.go [hunk 3]
```diff
@@ -143,27 +143,27 @@ func (i *invoiceRegistry) invoiceEventNotifier() {
...
-				case !event.isSettle &&
+				case event.newState == channeldb.ContractOpen &&
 					client.addIndex >= invoice.AddIndex:
 					continue
 
 				// These two states should never happen, but we
 				// log them just in case so we can detect this
 				// instance.
-				case !event.isSettle &&
+				case event.newState == channeldb.ContractOpen &&
```
Comment (halseth) [inline invoiceregistry.go]: Same question whether this should be `!= ContractSetteld` instead.

File: invoiceregistry.go [hunk 4]
```diff
@@ -89,7 +89,7 @@ func (i *invoiceRegistry) Stop() {
 // Only two event types are currently supported: newly created invoices, and
 // instance where invoices are settled.
 type invoiceEvent struct {
-	isSettle bool
+	newState channeldb.ContractState
```
Comment (halseth) [inline invoiceregistry.go]: nit: just name `state`?

File: invoiceregistry.go [hunk 5]
```diff
@@ -187,9 +187,10 @@ func (i *invoiceRegistry) invoiceEventNotifier() {
 				// don't send a notification twice, which can
 				// happen if a new event is added while we're
 				// catching up a new client.
-				if event.isSettle {
+				switch event.newState {
+				case channeldb.ContractSettled:
 					client.settleIndex = invoice.SettleIndex
-				} else {
+				case channeldb.ContractOpen:
```
Comment (halseth) [inline invoiceregistry.go]: default case for unknown state

File: invoiceregistry.go [hunk 6]
```diff
@@ -483,8 +486,11 @@ func (i *invoiceRegistry) SubscribeNotifications(addIndex, settleIndex uint64) *
 			case ntfn := <-client.ntfnQueue.ChanOut():
 				invoiceEvent := ntfn.(*invoiceEvent)
 
-				targetChan := client.NewInvoices
-				if invoiceEvent.isSettle {
+				var targetChan chan *channeldb.Invoice
+				switch invoiceEvent.newState {
+				case channeldb.ContractOpen:
+					targetChan = client.NewInvoices
+				case channeldb.ContractSettled:
```
Comment (halseth) [inline invoiceregistry.go]: default case

File: htlcswitch/link.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/link.go]: Yes. In this PR both are equivalent, but I am inverting the condition as a preparation for the hold invoice.

Comment (joostjager) [reply htlcswitch/link.go]: Same for comment below

**Comment by joostjager** [comment]
comments processed

**Comment by joostjager** [comment]
also added `String()` for `ContractState` type

File: channeldb/invoices.go [hunk 7]
```diff
@@ -74,6 +74,30 @@ const (
...
+// ContractState describes the state the invoice is in.
+type ContractState uint8
+
+const (
+	// ContractOpen means the invoice has only been created.
+	ContractOpen ContractState = 0
+
+	// ContractSettled means the htlc is settled and the invoice has been
+	// paid.
+	ContractSettled = 1
```
Comment (Roasbeef) [inline channeldb/invoices.go]: I think this one also needs to be explicitly specified as the prior one isn't using `iota`?

Comment (joostjager) [reply channeldb/invoices.go]: Fixed

**Comment by Roasbeef** [comment]
Ready to land after sign+rebase! (may need to sync with some of the other PRs that also modify `lnrpc` protos)

**Comment by joostjager** [comment]
Ptal @Roasbeef

**Pull request merged**